loader: dump datapath config object to disk

rgo3 opened the pull request:
For better debugging information, this commit ensures that the configuration objects with load-time config values will be written out to the state directory. With this change, sysdumps should automatically include these configuration objects that define how the datapath was configured.

Fixes: #41746

rgo3 committed:
loader: dump datapath config object to disk

For better debugging information, this commit ensures that the
configuration objects with load-time config values will be written out
to the state directory. With this change, sysdumps should automatically
include these configuration objects that define how the datapath was
configured.

Fixes: #41746

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

ti-mo commented on pkg/bpf/collection.go:
```diff
@@ -287,6 +296,31 @@ func renameMaps(coll *ebpf.CollectionSpec, renames map[string]string) error {
 	return nil
 }
 
+func serializeConstants(logger *slog.Logger, opts *CollectionOptions) {
```
I would return errors instead of logging and quietly continuing. CI jobs fail on `level=error`, but in a user's env, this likely won't raise alarm bells. A failure to write to the statedir is an indication that something else is afoul; we shouldn't happily continue in that case.

ti-mo commented on pkg/datapath/loader/base.go:
```diff
@@ -227,7 +227,8 @@ func (l *loader) reinitializeIPSec(lnc *datapath.LocalNodeConfiguration) error {
 		},
-		Constants: config.NewBPFNetwork(config.NodeConfig(lnc)),
+		Constants:  config.NewBPFNetwork(config.NodeConfig(lnc)),
+		ConfigPath: filepath.Join(option.Config.StateDir, "datapath_config", "bpf_encryption_network.cfg"),
```
I'd keep the naming of the .c/.o file and its corresponding .json aligned.

Also, see the constants at the top of `loader/compile.go`. We could add more consts for config files for individual bpf objects.

ti-mo reviewed (commented):
I like the directory structure we came up with for bpffs:
`/sys/fs/bpf/cilium/devices/<dev>/`. Could we mirror this for the statedir?

For example, config would go into `/sys/fs/bpf/cilium/devices/enp2s0/bpf_host.json`. I'm not sure where verifier logs go, but they could also be written here.

rgo3 committed:
loader: cleanup stale datapath config directories

Add cleanup logic to remove stale datapath configuration directories
during agent reinitialization. This handles cases where:

- IPSec is toggled off (removes bpf/bpf_network.json)
- Wireguard is toggled off (removes cilium_wg0 configs)
- Tunnel mode changes (removes overlay device configs)
- Devices are removed from the system

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
logfields: rename BPFSPath to BPFFSPath

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: fix printing constants in loader debug log

The change to support []any in CollectionOptions.Constants unfortunately broke
printing structs in the debug log since fmt doesn't step into slices to
pretty-print them: `constants="[]interface {}{(*config.BPFHost)(0xc002a46320)}"`.

This commit unwraps []any and prints each object individually:
`constants="[&config.BPFHost{AllowICMPFragNeeded:true, DeviceMTU:0x5dc, ...`

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: move applyConstants to new constants.go

Signed-off-by: Timo Beckers <timo@isovalent.com>
